### PR TITLE
Add dex provider (DFI-BTC)

### PR DESCRIPTION
**Pull request by monstrobishi**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

At the moment only mapping DFI-BTC pool as we can get other crypto info from various sources, Defichain DEX DFI-BTC is the best source for stable DFI pricing.

#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

File: packages/salmon-provider-dex/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      timestamp: new BigNumber(Date.now())
+    }
+  }
+
+  public async prices (symbols: string[]): Promise<AssetPrice[]> {
+    const response = await fetch(DEX_URL, {
+      method: 'GET'
+    })
+
+    const text = await response.text()
```
Comment (fuxingloh) [inline packages/salmon-provider-dex/src/index.ts:42]: https://github.com/DeFiCh/whale/blob/main/packages/whale-api-client/src/api/poolpair.ts

use this client instead?

File: packages/salmon-provider-dex/src/mapping.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import BigNumber from 'bignumber.js'
+import fetch from 'node-fetch'
+import {
+  JellyfishJSON
+} from '@defichain/jellyfish-json'
+
+interface DefichainSymbolMapping {
+  ticker: string
+  inverse: boolean
+  affector?: () => Promise<BigNumber>
```
Comment (fuxingloh) [inline packages/salmon-provider-dex/src/mapping.ts:10]: should rename this function, affector doesn't make sense?

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/salmon-provider-dex/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  public async prices (symbols: string[]): Promise<AssetPrice[]> {
+    const env: EnvironmentConfig = await getEnvironmentConfig()
+    const whaleClient = new WhaleApiClient({
+      url: env.oceanUrl,
+      network: env.network
+    })
+
+    const json = await whaleClient.poolpair.list(1000)
```
Comment (fuxingloh) [inline packages/salmon-provider-dex/src/index.ts:48]:
```suggestion
    const json = await whaleClient.poolpair.list(1000)
```

whale client doesn't allow calls like this, need to do pagination chaining to get all the results.

Comment (fuxingloh) [reply packages/salmon-provider-dex/src/index.ts:48]: For context: https://github.com/DeFiCh/whale/issues/268

Pagination chaining: https://github.com/DeFiCh/whale/blob/abbecbac7212556af71a4405241cca38d447163e/packages/whale-api-client/__tests__/api/address.test.ts#L144

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/salmon-provider-dex/src/index.ts (hunk 3 above)
Comment (fuxingloh) [reply packages/salmon-provider-dex/src/index.ts:48]: Protecting against future regression

**Review by jingyi2811 — APPROVED** [review]

File: packages/salmon-provider-dex/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  public async prices (symbols: string[]): Promise<AssetPrice[]> {
+    const env: EnvironmentConfig = await getEnvironmentConfig()
+    const whaleClient = new WhaleApiClient({
+      url: env.oceanUrl,
+      network: env.network
+    })
+
+    let json: any[] = []
+    let whaleResponse: ApiPagedResponse<any> = await whaleClient.poolpair.list(50)
```
Comment (fuxingloh) [inline packages/salmon-provider-dex/src/index.ts:49]: without enforcing the types, you are taking typescript soundness out of the equation

**Review by fuxingloh — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

File: packages/salmon-provider-dex/src/index.ts (hunk 4 above)
Comment (monstrobishi) [reply packages/salmon-provider-dex/src/index.ts:49]: I think whale client doesn't expose the poolpair structure type

Comment (fuxingloh) [reply packages/salmon-provider-dex/src/index.ts:49]: https://github.com/DeFiCh/wallet/blob/31ecd6da37a55b144b2e7f02974ec9102f7d3699/app/screens/AppNavigator/screens/Dex/DexScreen.tsx#L2

Comment (monstrobishi) [reply packages/salmon-provider-dex/src/index.ts:49]: I stand corrected! I am hesitant to import from dist dir though, I think interface types should be exposed at the root level

Comment (fuxingloh) [reply packages/salmon-provider-dex/src/index.ts:49]:
```ts
import { poolpairs } from '@defichain/whale-api-client'

const data: poolpairs.PoolPairData[] = []
```

I believe you can do this also

**Pull request merged**